### ISSUE TITLE
Example.cpp: Memory usage grows out of control if run with longer video

### ISSUE THREAD

**Issue by ferdnyc** [OpenShot/libopenshot#239]
@jonoomph — If you have a moment, a quick question about the libopenshot example code.

Interested in getting a better picture of when, where, and how libopenshot allocates memory during the encoding process (see OpenShot/openshot-qt#2811), I modified `src/examples/Example.cpp` with useful local paths, different codecs, a longer output stream (1000 frames instead of 100), and the correct hardware decoder type for my Nvidia GTX 710 GPU (VDPAU decoder), but no other substantial changes. Here's a complete diff from the checked-in source to my version:
```diff
diff --git a/src/examples/Example.cpp b/src/examples/Example.cpp
index 8033968..e4f3305 100644
--- a/src/examples/Example.cpp
+++ b/src/examples/Example.cpp
@@ -37,19 +37,19 @@ using namespace openshot;
 int main(int argc, char* argv[]) {
 
 	Settings *s = Settings::Instance();
-	s->HARDWARE_DECODER = 2; // 1 VA-API, 2 NVDEC
+	s->HARDWARE_DECODER = 6; // 1 VA-API, 2 NVDEC, 6 VDPAU, 7 QSV
 	s->HW_DE_DEVICE_SET = 1;
 
-    FFmpegReader r9("/home/jonathan/Videos/sintel_trailer-720p.mp4");
+    FFmpegReader r9("/home/ferd/rpmbuild/REPOS/openshot-qt/bugtesting/memory-tests/BBB-2m-clip1_00:00:00-00:02:02.mp4");
     r9.Open();
     r9.DisplayInfo();
 
     /* WRITER ---------------- */
-    FFmpegWriter w9("/home/jonathan/metadata.mp4");
+    FFmpegWriter w9("/var/tmp/example-out.mp4");
 
     // Set options
-    w9.SetAudioOptions(true, "libmp3lame", r9.info.sample_rate, r9.info.channels, r9.info.channel_layout, 128000);
-    w9.SetVideoOptions(true, "libx264", r9.info.fps, 1024, 576, Fraction(1,1), false, false, 3000000);
+    w9.SetAudioOptions(true, "aac", r9.info.sample_rate, r9.info.channels, r9.info.channel_layout, 128000);
+    w9.SetVideoOptions(true, "libx264", r9.info.fps, 1920, 1080, Fraction(1,1), false, false, 3000000);
 
     w9.info.metadata["title"] = "testtest";
 	w9.info.metadata["artist"] = "aaa";
@@ -63,7 +63,7 @@ int main(int argc, char* argv[]) {
     // Open writer
     w9.Open();
 
-    for (long int frame = 1; frame <= 100; frame++)
+    for (long int frame = 1; frame <= 1000; frame++)
     {
         //int frame_number = (rand() % 750) + 1;
         int frame_number = frame;
@@ -80,4 +80,4 @@ int main(int argc, char* argv[]) {
 	cout << "Completed successfully!" << endl;
 
     return 0;
-}
\ No newline at end of file
+}
```

I built `libopenshot.so` and `openshot-example` using CMake, then ran `src/openshot-example` from the build directory.

I was surprised to discover that the process's memory usage balloons out of control, growing continuously until all of my machine's 6GB of RAM is exhausted or I kill the process.

Is that the expected behavior of the `Example.cpp` code? Is it missing necessary cleanup code that would be required to keep the process size under control?

**Comment by jeffski** [comment]
Is this a GPU hardware issue or an issue in general with libopenshot. I've also had problems with libopenshot running out of memory especially when working with 1080 footage and applying effects and text. Just curious if you don't use hardware encoding do you get the same issue?

**Comment by ferdnyc** [comment]
@jeffski No hardware encoding in use, only the hardware _decoder_ is enabled. 

If I disable it (set `s->HARDWARE_DECODER = 0;`), the memory growth is **slower** by far. In fact, the process can successfully get through 1000 frames without exhausting my system RAM, only growing to around 2.5G before it finishes writing the output. But if I raise the frame count from 1000 to 3000, then the growth still continues unchecked until all RAM is allocated.

**Comment by ferdnyc** [comment]
Fixed by @ckirmse in #240
